cmd/kubeadm/app/phases/addons/addons_test.go

```diff
@@ -86,3 +92,69 @@ func TestCompileManifests(t *testing.T) {
 		}
 	}
 }
+
+// This unit test exists because tolerations weren't decoded correctly by kuberuntime.DecodeInto
+// In case that regression ever happens again, this unit test will catch it
+func TestTolerationsPreserved(t *testing.T) {
+	var tests = []struct {
+		manifest string
+		data     interface{}
+		f        func(data []byte) []v1.Toleration
+		expected []v1.Toleration
+	}{
+		{
+			manifest: KubeProxyDaemonSet,
+			data: struct{ Image, ClusterCIDR, MasterTaintKey string }{
+				Image:          "foo",
+				ClusterCIDR:    "- foo",
+				MasterTaintKey: kubeadmconstants.LabelNodeRoleMaster,
+			},
+			f: func(data []byte) []v1.Toleration {
+				kubeproxyDaemonSet := &extensions.DaemonSet{}
+				if err := kuberuntime.DecodeInto(api.Codecs.UniversalDecoder(), data, kubeproxyDaemonSet); err != nil {
```

Comment:
I checked the code of the Universal decoder, it will try to unmarhal and then convert the object to the internal version. I still don't understand how it's working. It's not relevant now since the test is removed. Please cc me next time kubeadm use the api.Codecs.